### PR TITLE
feat: add chrome and electron support

### LINKED ISSUE #254: Trying to install stable chrome version
I'm trying to install a stable chrome version for https://github.com/thepracticaldev/dev.to/blob/master/.gitpod.dockerfile

We need chrome in there because some of the unit tests use chrome headlessly.  

However, I'm struggling to add chrome in a dockerfile, since I believe chrome needs sudo permissions. Any idea how to accomplish this?

Thanks!

### LINKED ISSUE #515: Make Puppeteer work in default workspace
Coming here from: https://community.gitpod.io/t/using-puppeteer-libnss3-no-file-or-directory-issue-on-gitpod/1762?u=danielhabenicht

Where should I implement the solution? 
In the `full` or `base` image?

### PR DESCRIPTION
**Pull request by gengjiawen**
close https://github.com/gitpod-io/workspace-images/issues/457 
close #254
close #515

### CODE REVIEW

**Comment by gengjiawen** [comment]
Another solution is that we put chrome in full image.

**Review by jankeromnes — COMMENTED** [review]
Hey, awesome! Thank you for building an image specifically for front-end development.

I'm not fundamentally against having this image among the official Gitpod workspace images. However, I wonder if this is the most useful thing we can do, or if there are other potentially more effective ways to help front-end developers use Gitpod.

Also, on a general note, I believe that "simple" images (e.g. where you install one or two things without a lot of configuration) should rather be in documentation, or in reusable template repositories that users can simply clone and adapt. Images in this repository should be for more complicated setups, e.g. when it's not easy to install a tool and get it to work well with Gitpod (e.g. databases, which need to have their datasource reconfigured, and they need to be automatically started and kept running in Gitpod). In these cases, the Dockerfiles become very long and complicated, and it makes more sense to have them here.

So, in short, I'd be happy to merge this as is, but:

- @ghuntley Maybe this would make more sense as an [official Gitpod template repository](https://github.com/gitpod-io?q=template&type=&language=&sort=)?

- @gengjiawen Maybe it would be useful to install a few additional tools, or explain how to install & use them? E.g. it would be cool to have Cypress, Electron, etc. already pre-installed in this image (we could ship all the most common front-end tools for convenience, so that front-end developers can just use the image without installing or configuring anything new)

**Comment by gengjiawen** [comment]
> E.g. it would be cool to have Cypress, Electron

They generally in package.json and use different versions. It's very hard to customize.

**Comment by gengjiawen** [comment]
Found another usecase, wasm test require headless browser. I now use `gitpod/workspace-wasm:latest` to do related dev. So make more sense add chrome to full.

I have refactor my code.

**Comment by jankeromnes** [comment]
Cool, many thanks!

Notes (mostly for myself when I'll review better next week):
:
- It would be cool to expand full-vnc's README.md to explain what the image is about (and not just talk about Electron)
- We should confirm whether installing Chrome installs a graphical stack (X server) in full, which we don't want because it uses a lot of disk space

**Comment by gengjiawen** [comment]
> We should confirm whether installing Chrome installs a graphical stack

From what I test, it only increase less than 200M in Image size.

**Comment by gengjiawen** [comment]
ping @jankeromnes

**Comment by gengjiawen** [comment]
@jankeromnes Any update ?

**Comment by jankeromnes** [comment]
Could someone from @gitpod-io/engineering-workspace please take a look at these changes? 🙏

**Comment by gengjiawen** [comment]
I have pushed `gengjiawen/gitpod-fe` if anyone want to try.

**Comment by gengjiawen** [comment]
@princerachit Can you review this ? thx

**Comment by DanielHabenicht** [comment]
Just my two cents, maybe it is a clever idea to split the changes this way:

Add the base dependencies for electron (and puppeteer, so simply read "any automated frontend testing tool") to the base image of gitpod? This would only include: 

```dockerfile
RUN sudo apt-get update \
 && sudo apt-get install -y \
  libasound2-dev \
  libgtk-3-dev \
  libnss3-dev
```


Add the chrome browser to the image as described in this PR. 


This would make life easier for any frontend developer as they won't have to patch their gitpod image just to run the test.

**Comment by princerachit** [comment]
Hi @gengjiawen , As this is a frontend related change I think it would be better if it goes along with a frontend image like `node(-xxx)`.

**Comment by princerachit** [comment]
Also, please add tests. You can refer to the [contribution guide](https://github.com/gitpod-io/workspace-images/blob/master/CONTRIBUTING.md) to learn more.

**Comment by gengjiawen** [comment]
> Hi @gengjiawen , As this is a frontend related change I think it would be better if it goes along with a frontend image like `node(-xxx)`.

You mean put it in `chunks/lang-node/Dockerfile` ?

**Comment by gengjiawen** [comment]
@kylos101 PTAL

File: chunks/tool-chrome/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
+ENV TRIGGER_REBUILD=1
+
+# chrome and basic render font
+RUN apt-get update \
+  && apt-get install -y apt-transport-https \
+  && curl -sSL https://dl.google.com/linux/linux_signing_key.pub | sudo apt-key add - \
+  && echo "deb [arch=amd64] https://dl.google.com/linux/chrome/deb/ stable main" > /etc/apt/sources.list.d/google-chrome.list \
+  && apt-get update && apt-get install -y google-chrome-stable --no-install-recommends \
+  && apt-get install -y fonts-noto fonts-noto-cjk
```
Comment (kylos101) [inline chunks/tool-chrome/Dockerfile:15]: May I ask you to share the use case for installing these fonts? For example, what are they a dependency of?

File: chunks/tool-chrome/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+# Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
+ENV TRIGGER_REBUILD=1
+
+# chrome and basic render font
+RUN apt-get update \
+  && apt-get install -y apt-transport-https \
+  && curl -sSL https://dl.google.com/linux/linux_signing_key.pub | sudo apt-key add - \
+  && echo "deb [arch=amd64] https://dl.google.com/linux/chrome/deb/ stable main" > /etc/apt/sources.list.d/google-chrome.list \
+  && apt-get update && apt-get install -y google-chrome-stable --no-install-recommends \
```
Comment (kylos101) [inline chunks/tool-chrome/Dockerfile:14]: Awesome, we are super close now! Let me know what you think of the below suggestion? This will help this new chunk follow [our contribution guide](https://github.com/gitpod-io/workspace-images/blob/master/CONTRIBUTING.md#adding-a-chunk). I think you'll like number 8 ([here is supporting info](https://www.linuxuprising.com/2021/01/apt-key-is-deprecated-how-to-add.html)).

And, now that we're socializing this PR, we should add a number 9 (when and why to use `install-packages`).

For example, we're using apt 2.0, and it looks like [https is supported out of the box with 1.5+](https://unix.stackexchange.com/a/90229).

```shell
USER root
RUN curl -sSL https://dl.google.com/linux/linux_signing_key.pub | sudo gpg --dearmor -o /usr/share/keyrings/chrome-archive-keyring.gpg \
   && echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/chrome-archive-keyring.gpg] https://dl.google.com/linux/chrome/deb/ \
    stable main" | sudo tee /etc/apt/sources.list.d/chrome.list > /dev/null \
    && apt update

USER gitpod
# you can append more package you with to install
# install-packages does many things for you, like update and clean-up afterwards
sudo install-packages google-chrome-stable

```

File: chunks/tool-chrome/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  && apt-get install -y fonts-noto fonts-noto-cjk 
+
+# misc deps for electron and puppeteer to run
+RUN sudo apt-get update \
+ && sudo apt-get install -y \
+  libasound2-dev \
+  libgtk-3-dev \
+  libnss3-dev
```
Comment (kylos101) [inline chunks/tool-chrome/Dockerfile:22]: @gengjiawen may I ask you to share references for these dependencies? I ask because a searched some results, the the libraries needed appear to vary. [[1](https://postsrc.com/code-snippets/how-to-install-latest-stable-chrome-version-in-ubuntu-puppeteer)][[2](https://www.electronjs.org/docs/latest/development/build-instructions-linux)]

If you were able to install these manually and get them working from our current node image, perhaps you could share a short Loom video?

**Review by kylos101 — COMMENTED** [review]
@gengjiawen we really appreciate you taking the time to work with us, fixing these issues, and organizing this PR. :100: :muscle: 

Take a peek at my comments and let me know what you think? :+1:

File: chunks/tool-chrome/Dockerfile (hunk 1 above)
Comment (gengjiawen) [reply chunks/tool-chrome/Dockerfile:15]: when chrome render page to pdf, it can cause unknown character. (Like gitbook render to pdf)

File: dazzle.yaml [hunk 4]
```diff
@@ -28,6 +28,7 @@ combiner:
         - lang-ruby:2.7
         - lang-rust
         - tool-brew
+        - tool-chrome
```
[the remark this answers is not in the document]
Comment (gengjiawen) [reply dazzle.yaml:31]: I am pretty like it in full. I can remove it since we can add it later. 

Can you do a poll on twitter so we can get further user feedback ?

File: chunks/tool-chrome/Dockerfile (hunk 3 above)
Comment (gengjiawen) [reply chunks/tool-chrome/Dockerfile:22]: This is what I find when I use gitpod for electron dev: see https://github.com/gengjiawen/electron-devdocs/blob/1937127c45a8d138400f68b3022b4280170bf6db/.gitpod.Dockerfile#L3.

File: chunks/tool-chrome/Dockerfile (hunk 2 above)
Comment (gengjiawen) [reply chunks/tool-chrome/Dockerfile]: after apply this changes, when I run `apt update`, I got 

```console
E: Conflicting values set for option Signed-By regarding source https://dl.google.com/linux/chrome/deb/ stable: /usr/share/keyrings/chrome-archive-keyring.gpg != 
E: The list of sources could not be read.
```

File: dazzle.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (DanielHabenicht) [reply dazzle.yaml]: As I already mentioned above, how about adding only the base dependencies to the full (or even base image)?
In most packages or tests the needed chrome version will be downloaded anyway, during e.g. package install. 
And just installing the 4 apks should't make it much larger. 
Also chrome for testing is used by many languages for any e2e tests.

File: chunks/tool-chrome/Dockerfile (hunk 2 above)
Comment (gengjiawen) [reply chunks/tool-chrome/Dockerfile]: @kylos101

Comment (kylos101) [reply chunks/tool-chrome/Dockerfile]: Hey @gengjiawen ,

This works better (I just tested).

```dockerfile
USER root

# Dazzle does not rebuild a layer until one of its lines are changed. Increase this counter to rebuild this layer.
ENV TRIGGER_REBUILD=1

# chrome and basic render font
RUN curl -fsSL https://dl.google.com/linux/linux_signing_key.pub | sudo gpg --dearmor -o /usr/share/keyrings/chrome-archive-keyring.gpg \
   && echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/chrome-archive-keyring.gpg] https://dl.google.com/linux/chrome/deb/ stable main" | sudo tee /etc/apt/sources.list.d/chrome.list > /dev/null \
    && apt update \
    && install-packages google-chrome-stable

USER gitpod

# misc deps for electron and puppeteer to run
RUN sudo install-packages \
    libasound2-dev libgtk-3-dev libnss3-dev \
    fonts-noto fonts-noto-cjk
```

Can you check, to assert it fits your needs?

**Comment by kylos101** [comment]
Hey @gengjiawen it looks like you pushed another fix in [6a8540c](https://github.com/gitpod-io/workspace-images/pull/471/commits/6a8540cca4f091f1cf518f0d03bbe7a2c0802376), let us know when ready to review again? :wave:

**Comment by princerachit** [comment]
@kylos101 can you work with @gengjiawen to get this to completion? There has been more changes in the master branch so rebasing is required too.

**Comment by kylos101** [comment]
Hey @gengjiawen , I believe I fixed the issue causing test failure, [the build is running here](https://github.com/gitpod-io/workspace-images/runs/5426578349?check_suite_focus=true). If it's green, we'll be able to approve this PR.

**Comment by kylos101** [comment]
Hey @gengjiawen the tests passed. I just rebased this feature branch with the upstream master. Assuming that is successful too, I'll squash and merge this PR. And then we'll finally have support for the `tool-chrome` chunk in the node workspace images. 

Appreciate your help and patience.

**Review by kylos101 — APPROVED** [review]
LGTM

**Comment by kylos101** [comment]
Hey @gengjiawen , once [this finishes](https://github.com/gitpod-io/workspace-images/actions/runs/1935842896), the new image will be available on `workspace-node` and `workspace-node-lts`, and have a fresh timestamp tag (YYYY-MM-DD-hh-mm-ss). `latest` won't be updated for ~1 week, so you'll want to be on the lookout for the new tag.

**Comment by kylos101** [comment]
@gengjiawen here is [the image](https://hub.docker.com/layers/gitpod/workspace-node-lts/2022-03-04-21-39-01/images/sha256-00405feb6eb6926045e50652cc00c46de1a09e5d31dfb0370cb2dd10dcb60eb5?context=explore), have a nice weekend! :rocket:

File: dazzle.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (gengjiawen) [reply dazzle.yaml]: > As I already mentioned above, how about adding only the base dependencies to the full (or even base image)? In most packages or tests the needed chrome version will be downloaded anyway, during e.g. package install. And just installing the 4 apks should't make it much larger. Also chrome for testing is used by many more languages (then node/javascript) for any e2e tests. @kylos101

I feel really unconveninent without it in full image. I checked the size, it's only around 130m increase.
It should be in full in the long run.
@kylos101

[the remark this answers is not in the document]
Comment (gengjiawen) [reply dazzle.yaml]: PR: https://github.com/gitpod-io/workspace-images/pull/903
